File: typescript/packages/mechanisms/evm/src/batch-settlement/server/managed.ts
import type {
  SettleContext,
  SettleResultContext,
  VerifiedPaymentCanceledContext,
  VerifyContext,
  VerifyResultContext,
} from "@x402/core/server";
import type { PaymentRequirements, SchemePaymentRequiredContext } from "@x402/core/types";
import { isBatchSettlementPayload, isBatchSettlementRefundPayload } from "../types";
import type { BatchSettlementChannelStateExtra, BatchSettlementVoucherStateExtra } from "../types";
import { BATCH_SETTLEMENT_SCHEME } from "../constants";
import * as Errors from "../errors";
import type { Channel } from "./storage";
import {
  abortIfBelowMinDeposit,
  abortIfChannelUnbound,
  abortIfUnexpectedServerAuthoredSettleFields,
  skipHandlerForRefund,
  writeCorrectiveAcceptExtra,
} from "./verify";
import { buildRefundSettlementFields } from "./settle";
import { readChannelStateExtra, readExtraNumber, readExtraString } from "./utils";
import type { BatchSettlementEvmScheme } from "./scheme";

/**
 * Pass-through verify: min-deposit + channel-id binding only.
 *
 * @param scheme - Owning scheme.
 * @param ctx - Verify lifecycle context.
 * @returns Abort on binding/min-deposit failure; otherwise continue to facilitator `/verify`.
 */
export async function handleManagedBeforeVerify(
  scheme: BatchSettlementEvmScheme,
  ctx: VerifyContext,
): Promise<void | { abort: true; reason: string; message?: string }> {
  const { paymentPayload, requirements } = ctx;
  const raw = paymentPayload.payload;
  if (!isBatchSettlementPayload(raw)) {
    return;
  }

  const serverFieldAbort = abortIfUnexpectedServerAuthoredSettleFields(raw);
  if (serverFieldAbort) {
    return serverFieldAbort;
  }

  const minDepositAbort = await abortIfBelowMinDeposit(scheme, raw, requirements);
  if (minDepositAbort) {
    return minDepositAbort;
  }

  return abortIfChannelUnbound(raw, requirements.network);
}

/**
 * True when a verify rejection carries a resyncable cumulative baseline.
 * The facilitator emits `ErrCumulativeAmountMismatch` for managed
 * voucher-store drift, while the shared client handshake also accepts
 * `ErrCumulativeAmountBelowClaimed`; the server must propagate corrective
 * extras for both or the client falls back to stale onchain recovery.
 *
 * @param reason - Facilitator verify rejection reason, when present.
 * @returns Whether the server should attach corrective cumulative baseline extras.
 */
function isCorrectiveMismatch(reason: string | undefined): boolean {
  return (
    reason === Errors.ErrCumulativeAmountMismatch ||
    reason === Errors.ErrCumulativeAmountBelowClaimed
  );
}

/**
 * After facilitator verify: stash a channel view for refund enrichment / replica,
 * or stash corrective extras. Refunds skip the resource handler.
 *
 * @param scheme - Owning scheme.
 * @param ctx - Post-verify context.
 * @returns `skipHandler` for refunds; otherwise void.
 */
export async function handleManagedAfterVerify(
  scheme: BatchSettlementEvmScheme,
  ctx: VerifyResultContext,
): Promise<
  | void
  | { skipHandler: true; response?: { contentType?: string; body?: unknown } }
  | { abort: true; reason: string; message?: string }
> {
  const { paymentPayload, result } = ctx;
  const raw = paymentPayload.payload;
  if (!isBatchSettlementPayload(raw)) {
    return;
  }

  if (!result.isValid) {
    if (isCorrectiveMismatch(result.invalidReason)) {
      const ex = result.extra ?? {};
      const channelState = readCorrectiveChannelState(ex);
      const voucherState = readCorrectiveVoucherState(ex);
      if (channelState || voucherState) {
        scheme.mergeRequestContext(paymentPayload, {
          correctiveChannelState: channelState,
          correctiveVoucherState: voucherState,
        });
      }
    }
    return;
  }

  const ex = result.extra ?? {};
  const now = Date.now();
  const channelSnapshot: Channel = {
    channelId: raw.voucher.channelId,
    channelConfig: raw.channelConfig,
    chargedCumulativeAmount: readExtraString(ex, "chargedCumulativeAmount", "0"),
    signedMaxClaimable: raw.voucher.maxClaimableAmount,
    signature: raw.voucher.signature,
    balance: readExtraString(ex, "balance", "0"),
    totalClaimed: readExtraString(ex, "totalClaimed", "0"),
    withdrawRequestedAt: readExtraNumber(ex, "withdrawRequestedAt", 0),
    refundNonce: readExtraNumber(ex, "refundNonce", 0),
    lastRequestTimestamp: now,
  };

  const pendingId = typeof ex.pendingId === "string" && ex.pendingId ? ex.pendingId : undefined;

  scheme.mergeRequestContext(paymentPayload, {
    channelId: raw.voucher.channelId,
    channelSnapshot,
    ...(pendingId ? { pendingId, reservationCommitted: true } : {}),
  });

  if (isBatchSettlementRefundPayload(raw)) {
    return skipHandlerForRefund(raw.voucher.channelId);
  }
}

/**
 * Copies facilitator-supplied corrective extras onto the matching 402 accept.
 *
 * @param scheme - Owning scheme.
 * @param ctx - Payment-required response context.
 */
export async function handleManagedEnrichPaymentRequiredResponse(
  scheme: BatchSettlementEvmScheme,
  ctx: SchemePaymentRequiredContext,
): Promise<void> {
  if (!isCorrectiveMismatch(ctx.error) || !ctx.paymentPayload) {
    return;
  }

  const requestContext = scheme.takeRequestContext(ctx.paymentPayload);
  const channelState = requestContext?.correctiveChannelState;
  const voucherState = requestContext?.correctiveVoucherState;
  if (!channelState || !voucherState) {
    return;
  }

  const accept = ctx.requirements.find(
    req =>
      req.scheme === BATCH_SETTLEMENT_SCHEME &&
      req.network === ctx.paymentPayload?.accepted.network,
  );
  if (!accept) {
    return;
  }

  writeCorrectiveAcceptExtra(accept, channelState, voucherState);
}

/** Managed mode has no local admission lock or settle short-circuit. */
async function noopManagedHook(): Promise<void> {}

export const handleManagedBeforeSettle = noopManagedHook;
export const handleManagedEnrichSettlementResponse = noopManagedHook;
export const handleManagedVerifyFailure = noopManagedHook;
export const handleManagedSettleFailure = noopManagedHook;
export const handleManagedVerifiedPaymentCanceled = noopManagedHook;

/**
 * Settles a cancel so the facilitator can drop the admission lock.
 * Enrichment stamps `cancel: true`; the facilitator must not charge, deposit,
 * or refund.
 *
 * @param ctx - Cancellation context from the resource server.
 * @returns Zero-amount requirements for batch-settlement payloads; void otherwise.
 */
export function handleManagedSettleOnCancel(
  ctx: VerifiedPaymentCanceledContext,
): PaymentRequirements | void {
  if (
    ctx.reason !== "handler_failed" &&
    ctx.reason !== "handler_threw" &&
    ctx.reason !== "after_verify_aborted"
  ) {
    return;
  }
  if (!isBatchSettlementPayload(ctx.paymentPayload.payload)) {
    return;
  }
  return { ...ctx.requirements, amount: "0" };
}

/**
 * Echoes verify `pendingId` onto `/settle` and completes a managed refund.
 * Cancel stamps `cancel: true` so the facilitator only releases the lock.
 * Never sets `claimAuthorizerSignature`.
 *
 * @param scheme - Owning scheme.
 * @param ctx - Settlement context.
 * @returns Additive settle fields, or nothing when there is nothing to attach.
 */
export async function handleManagedEnrichSettlementPayload(
  scheme: BatchSettlementEvmScheme,
  ctx: SettleContext,
): Promise<Record<string, unknown> | void> {
  const { paymentPayload, requirements } = ctx;
  const raw = paymentPayload.payload;
  if (!isBatchSettlementPayload(raw)) {
    return;
  }

  const pendingId = scheme.readRequestContext(paymentPayload)?.pendingId;
  const pendingFields = pendingId ? { pendingId } : {};

  if (ctx.phase === "cancel") {
    return { ...pendingFields, cancel: true };
  }

  if (isBatchSettlementRefundPayload(raw)) {
    const snapshot = scheme.readRequestContext(paymentPayload)?.channelSnapshot;
    if (!snapshot) {
      throw new Error(Errors.ErrMissingChannel);
    }

    return {
      ...(await buildRefundSettlementFields({
        channel: snapshot,
        raw,
        channelId: raw.voucher.channelId,
        network: requirements.network,
        refundSigner: scheme.getRefundAuthorizerSigner(),
        includeClaimAuthorizerSignature: false,
      })),
      ...pendingFields,
    };
  }

  return pendingId ? pendingFields : undefined;
}

/**
 * Replica upsert after a successful managed settle. Not read on the hot path.
 * Cancel settles leave the replica watermark unchanged, so this is a no-op.
 *
 * @param scheme - Owning scheme.
 * @param ctx - Post-settle context.
 */
export async function handleManagedAfterSettle(
  scheme: BatchSettlementEvmScheme,
  ctx: SettleResultContext,
): Promise<void> {
  const { paymentPayload, result } = ctx;
  if (!result.success || ctx.phase === "cancel") {
    return;
  }

  const raw = paymentPayload.payload;
  if (!isBatchSettlementPayload(raw)) {
    return;
  }

  const storage = scheme.getStorage();
  const channelId = raw.voucher.channelId;
  const channelState = readChannelStateExtra(result.extra);
  const now = Date.now();
  const charged =
    (channelState && readExtraString(channelState, "chargedCumulativeAmount", "")) ||
    scheme.readRequestContext(paymentPayload)?.channelSnapshot?.chargedCumulativeAmount ||
    "0";

  if (isBatchSettlementRefundPayload(raw)) {
    const balance = readExtraString(channelState, "balance", "0");
    if (BigInt(balance) <= BigInt(charged)) {
      await storage.updateChannel(channelId, current => (current ? undefined : current));
      return;
    }
  }

  await storage.updateChannel(channelId, current => {
    if (current && !isBatchSettlementRefundPayload(raw)) {
      try {
        if (BigInt(charged) < BigInt(current.chargedCumulativeAmount)) {
          return current;
        }
      } catch {
        // Non-numeric watermarks fall through to the normal upsert below.
      }
    }
    const base = current ?? scheme.readRequestContext(paymentPayload)?.channelSnapshot;
    return {
      channelId,
      channelConfig: raw.channelConfig,
      chargedCumulativeAmount: charged || base?.chargedCumulativeAmount || "0",
      signedMaxClaimable: raw.voucher.maxClaimableAmount,
      signature: raw.voucher.signature,
      balance: readExtraString(channelState, "balance", base?.balance ?? "0"),
      totalClaimed: readExtraString(channelState, "totalClaimed", base?.totalClaimed ?? "0"),
      withdrawRequestedAt: readExtraNumber(
        channelState,
        "withdrawRequestedAt",
        base?.withdrawRequestedAt ?? 0,
      ),
      refundNonce: readExtraNumber(channelState, "refundNonce", base?.refundNonce ?? 0),
      lastRequestTimestamp: now,
    };
  });
}

/**
 * Reads a corrective channel snapshot from facilitator verify extras.
 *
 * @param extra - Verify response extra.
 * @returns Typed channel state, or undefined.
 */
function readCorrectiveChannelState(
  extra: Record<string, unknown>,
): BatchSettlementChannelStateExtra | undefined {
  const value = extra.channelState;
  if (typeof value !== "object" || value === null) {
    return undefined;
  }
  return value as BatchSettlementChannelStateExtra;
}

/**
 * Reads a corrective voucher proof from facilitator verify extras.
 *
 * @param extra - Verify response extra.
 * @returns Typed voucher state, or undefined.
 */
function readCorrectiveVoucherState(
  extra: Record<string, unknown>,
): BatchSettlementVoucherStateExtra | undefined {
  const value = extra.voucherState;
  if (typeof value !== "object" || value === null) {
    return undefined;
  }
  return value as BatchSettlementVoucherStateExtra;
}

File: go/mechanisms/evm/batch-settlement/server/managed.go
package server

import (
	"errors"
	"math/big"
	"time"

	x402 "github.com/x402-foundation/x402/go/v2"
	batchsettlement "github.com/x402-foundation/x402/go/v2/mechanisms/evm/batch-settlement"
	"github.com/x402-foundation/x402/go/v2/types"
)

func isBatchSettlementPayload(raw map[string]interface{}) bool {
	return batchsettlement.IsDepositPayload(raw) ||
		batchsettlement.IsVoucherPayload(raw) ||
		batchsettlement.IsRefundPayload(raw)
}

// isCorrectiveMismatch reports whether a verify rejection carries a
// resyncable cumulative baseline. The facilitator emits
// ErrCumulativeAmountMismatch for managed voucher-store drift, while the
// shared client handshake also accepts ErrCumulativeBelowClaimed (canonical
// facilitator form); the server must propagate corrective extras for both or
// the client's ProcessCorrectivePaymentRequired sees a 402 without
// channelState/voucherState and falls back to stale onchain recovery.
func isCorrectiveMismatch(reason string) bool {
	return reason == batchsettlement.ErrCumulativeAmountMismatch ||
		reason == batchsettlement.ErrCumulativeBelowClaimed
}

// handleManagedBeforeVerify is a pass-through verify: min-deposit and
// channel-id binding only. Admission stays with the facilitator.
func handleManagedBeforeVerify(s *BatchSettlementEvmScheme, ctx x402.VerifyContext) (*x402.BeforeHookResult, error) {
	raw := ctx.Payload.GetPayload()
	if !isBatchSettlementPayload(raw) {
		return nil, nil
	}
	if abort := AbortIfUnexpectedServerAuthoredSettleFields(raw); abort != nil {
		return abort, nil
	}
	minDepositAbort, hintErr := AbortIfBelowMinDeposit(s, raw, ctx.Requirements)
	if hintErr != nil {
		return nil, hintErr
	}
	if minDepositAbort != nil {
		return minDepositAbort, nil
	}
	return AbortIfChannelUnbound(raw, ctx.Requirements.GetNetwork()), nil
}

// handleManagedAfterVerify stashes a channel view for refund enrichment and
// the replica, or corrective extras. Refunds skip the resource handler.
func handleManagedAfterVerify(s *BatchSettlementEvmScheme, ctx x402.VerifyResultContext) (*x402.AfterVerifyResult, error) {
	raw := ctx.Payload.GetPayload()
	if !isBatchSettlementPayload(raw) {
		return nil, nil
	}
	if ctx.Result == nil {
		return nil, nil
	}
	if !ctx.Result.IsValid {
		if isCorrectiveMismatch(ctx.Result.InvalidReason) {
			ex := ctx.Result.Extra
			channelState := readCorrectiveChannelState(ex)
			voucherState := readCorrectiveVoucherState(ex)
			if channelState != nil || voucherState != nil {
				s.MergeRequestContext(ctx.Payload, BatchSettlementRequestContext{
					CorrectiveChannelState: channelState,
					CorrectiveVoucherState: voucherState,
				})
			}
		}
		return nil, nil
	}

	ex := ctx.Result.Extra
	voucher, _ := raw["voucher"].(map[string]interface{})
	channelId, _ := voucher["channelId"].(string)
	maxClaimable, _ := voucher["maxClaimableAmount"].(string)
	signature, _ := voucher["signature"].(string)
	cfgMap, _ := raw["channelConfig"].(map[string]interface{})
	cfg, _ := batchsettlement.ChannelConfigFromMap(cfgMap)

	snapshot := &ChannelSession{
		ChannelId:               channelId,
		ChannelConfig:           cfg,
		ChargedCumulativeAmount: mapStringField(ex, "chargedCumulativeAmount", "0"),
		SignedMaxClaimable:      maxClaimable,
		Signature:               signature,
		Balance:                 mapStringField(ex, "balance", "0"),
		TotalClaimed:            mapStringField(ex, "totalClaimed", "0"),
		WithdrawRequestedAt:     mapIntField(ex, "withdrawRequestedAt", 0),
		RefundNonce:             mapIntField(ex, "refundNonce", 0),
		LastRequestTimestamp:    time.Now().UnixMilli(),
	}

	partial := BatchSettlementRequestContext{
		ChannelId:       channelId,
		ChannelSnapshot: snapshot,
	}
	if pendingId, ok := ex["pendingId"].(string); ok && pendingId != "" {
		partial.PendingId = pendingId
		partial.ReservationCommitted = reservationFlag(true)
	}
	s.MergeRequestContext(ctx.Payload, partial)

	if batchsettlement.IsRefundPayload(raw) {
		return SkipHandlerForRefund(channelId), nil
	}
	return nil, nil
}

// handleManagedEnrichPaymentRequiredResponse copies facilitator-supplied
// corrective extras onto the matching 402 accept.
func handleManagedEnrichPaymentRequiredResponse(s *BatchSettlementEvmScheme, ctx x402.PaymentRequiredContext) {
	if !isCorrectiveMismatch(ctx.Error) || ctx.PaymentPayload == nil {
		return
	}
	requestContext := s.TakeRequestContext(ctx.PaymentPayload)
	if requestContext == nil || requestContext.CorrectiveChannelState == nil || requestContext.CorrectiveVoucherState == nil {
		return
	}
	network := ctx.PaymentPayload.Accepted.Network
	for i := range ctx.Requirements {
		if ctx.Requirements[i].Scheme != batchsettlement.SchemeBatched {
			continue
		}
		if ctx.Requirements[i].Network != network {
			continue
		}
		WriteCorrectiveAcceptExtra(&ctx.Requirements[i], *requestContext.CorrectiveChannelState, *requestContext.CorrectiveVoucherState)
	}
}

func handleManagedBeforeSettle(*BatchSettlementEvmScheme, x402.SettleContext) (*x402.BeforeHookResult, error) {
	return nil, nil
}

func handleManagedEnrichSettlementResponse(*BatchSettlementEvmScheme, x402.SettleResultContext) (map[string]interface{}, error) {
	return nil, nil
}

func handleManagedVerifyFailure(*BatchSettlementEvmScheme, x402.VerifyFailureContext) (*x402.VerifyFailureHookResult, error) {
	return nil, nil
}

func handleManagedSettleFailure(*BatchSettlementEvmScheme, x402.SettleFailureContext) (*x402.SettleFailureHookResult, error) {
	return nil, nil
}

func handleManagedVerifiedPaymentCanceled(*BatchSettlementEvmScheme, x402.VerifiedPaymentCanceledContext) error {
	return nil
}

// handleManagedSettleOnCancel returns zero-amount requirements so the
// facilitator can drop the admission lock without charging, depositing, or refunding.
func handleManagedSettleOnCancel(ctx x402.VerifiedPaymentCanceledContext) (*types.PaymentRequirements, error) {
	switch ctx.Reason {
	case x402.CancellationReasonHandlerFailed, x402.CancellationReasonHandlerThrew, x402.CancellationReasonAfterVerifyAborted:
	default:
		return nil, nil
	}
	if ctx.Payload == nil || !isBatchSettlementPayload(ctx.Payload.GetPayload()) {
		return nil, nil
	}
	req := paymentRequirementsFromView(ctx.Requirements)
	req.Amount = "0"
	return &req, nil
}

// handleManagedEnrichSettlementPayload echoes verify pendingId onto settle and
// completes a managed refund. Cancel stamps cancel so the facilitator only
// releases the lock. Never sets claimAuthorizerSignature.
func handleManagedEnrichSettlementPayload(s *BatchSettlementEvmScheme, ctx x402.SettleContext) (map[string]interface{}, error) {
	raw := ctx.Payload.GetPayload()
	if !isBatchSettlementPayload(raw) {
		return nil, nil
	}

	var pendingId string
	if rc := s.ReadRequestContext(ctx.Payload); rc != nil {
		pendingId = rc.PendingId
	}
	pendingFields := map[string]interface{}{}
	if pendingId != "" {
		pendingFields["pendingId"] = pendingId
	}

	if ctx.Phase == x402.SettlePhaseCancel {
		out := map[string]interface{}{"cancel": true}
		for k, v := range pendingFields {
			out[k] = v
		}
		return out, nil
	}

	if batchsettlement.IsRefundPayload(raw) {
		rc := s.ReadRequestContext(ctx.Payload)
		if rc == nil || rc.ChannelSnapshot == nil {
			return nil, errors.New(batchsettlement.ErrMissingChannel)
		}
		voucher, _ := raw["voucher"].(map[string]interface{})
		amount, _ := raw["amount"].(string)
		maxClaimable, _ := voucher["maxClaimableAmount"].(string)
		signature, _ := voucher["signature"].(string)
		channelId, _ := voucher["channelId"].(string)
		fields, err := BuildRefundSettlementFields(RefundSettlementFields{
			Ctx:                             ctx.Ctx,
			Channel:                         rc.ChannelSnapshot,
			ChannelConfig:                   rc.ChannelSnapshot.ChannelConfig,
			MaxClaimableAmount:              maxClaimable,
			Signature:                       signature,
			Amount:                          amount,
			ChannelId:                       channelId,
			Network:                         ctx.Requirements.GetNetwork(),
			RefundSigner:                    s.GetRefundAuthorizerSigner(),
			IncludeClaimAuthorizerSignature: false,
		})
		if err != nil {
			return nil, err
		}
		for k, v := range pendingFields {
			fields[k] = v
		}
		return fields, nil
	}

	if pendingId == "" {
		return nil, nil
	}
	return pendingFields, nil
}

// handleManagedAfterSettle upserts the replica after a successful managed
// settle. Cancel settles leave the replica watermark unchanged.
func handleManagedAfterSettle(s *BatchSettlementEvmScheme, ctx x402.SettleResultContext) error {
	if ctx.Result == nil || !ctx.Result.Success || ctx.Phase == x402.SettlePhaseCancel {
		return nil
	}
	raw := ctx.Payload.GetPayload()
	if !isBatchSettlementPayload(raw) {
		return nil
	}

	voucher, _ := raw["voucher"].(map[string]interface{})
	channelId, _ := voucher["channelId"].(string)
	maxClaimable, _ := voucher["maxClaimableAmount"].(string)
	signature, _ := voucher["signature"].(string)
	cfgMap, _ := raw["channelConfig"].(map[string]interface{})
	cfg, _ := batchsettlement.ChannelConfigFromMap(cfgMap)

	channelState := readChannelStateFromExtra(ctx.Result.Extra)
	now := time.Now().UnixMilli()
	charged := ""
	if channelState != nil {
		charged = channelState.ChargedCumulativeAmount
	}
	if charged == "" {
		if rc := s.ReadRequestContext(ctx.Payload); rc != nil && rc.ChannelSnapshot != nil {
			charged = rc.ChannelSnapshot.ChargedCumulativeAmount
		}
	}
	if charged == "" {
		charged = "0"
	}

	if batchsettlement.IsRefundPayload(raw) {
		balance := "0"
		if channelState != nil && channelState.Balance != "" {
			balance = channelState.Balance
		}
		bal, _ := new(big.Int).SetString(balance, 10)
		chg, _ := new(big.Int).SetString(charged, 10)
		if bal == nil {
			bal = big.NewInt(0)
		}
		if chg == nil {
			chg = big.NewInt(0)
		}
		if bal.Cmp(chg) <= 0 {
			_, err := s.GetStorage().UpdateChannel(ctx.Ctx, channelId, func(current *ChannelSession) *ChannelSession {
				if current == nil {
					return current
				}
				return nil
			})
			return err
		}
	}

	_, err := s.GetStorage().UpdateChannel(ctx.Ctx, channelId, func(current *ChannelSession) *ChannelSession {
		if current != nil && !batchsettlement.IsRefundPayload(raw) {
			incoming, okIncoming := new(big.Int).SetString(charged, 10)
			stored, okStored := new(big.Int).SetString(current.ChargedCumulativeAmount, 10)
			if okIncoming && okStored && incoming.Cmp(stored) < 0 {
				return current
			}
		}
		var base *ChannelSession
		if current != nil {
			base = current
		} else if rc := s.ReadRequestContext(ctx.Payload); rc != nil {
			base = rc.ChannelSnapshot
		}
		next := ChannelSession{
			ChannelId:               channelId,
			ChannelConfig:           cfg,
			ChargedCumulativeAmount: charged,
			SignedMaxClaimable:      maxClaimable,
			Signature:               signature,
			Balance:                 "0",
			TotalClaimed:            "0",
			LastRequestTimestamp:    now,
		}
		if base != nil {
			if next.ChargedCumulativeAmount == "" {
				next.ChargedCumulativeAmount = base.ChargedCumulativeAmount
			}
			next.Balance = base.Balance
			next.TotalClaimed = base.TotalClaimed
			next.WithdrawRequestedAt = base.WithdrawRequestedAt
			next.RefundNonce = base.RefundNonce
		}
		if next.ChargedCumulativeAmount == "" {
			next.ChargedCumulativeAmount = "0"
		}
		if channelState != nil {
			if channelState.Balance != "" {
				next.Balance = channelState.Balance
			}
			if channelState.TotalClaimed != "" {
				next.TotalClaimed = channelState.TotalClaimed
			}
			next.WithdrawRequestedAt = channelState.WithdrawRequestedAt
			if channelState.RefundNonce != "" {
				if n, ok := new(big.Int).SetString(channelState.RefundNonce, 10); ok {
					next.RefundNonce = int(n.Int64())
				}
			}
		}
		return &next
	})
	return err
}

func readCorrectiveChannelState(extra map[string]interface{}) *batchsettlement.BatchSettlementChannelStateExtra {
	if extra == nil {
		return nil
	}
	raw, ok := extra["channelState"].(map[string]interface{})
	if !ok {
		return nil
	}
	out := &batchsettlement.BatchSettlementChannelStateExtra{}
	out.ChannelId, _ = raw["channelId"].(string)
	out.Balance, _ = raw["balance"].(string)
	out.TotalClaimed, _ = raw["totalClaimed"].(string)
	out.ChargedCumulativeAmount, _ = raw["chargedCumulativeAmount"].(string)
	if v, ok := raw["refundNonce"].(string); ok {
		out.RefundNonce = v
	} else if v, ok := raw["refundNonce"].(float64); ok {
		out.RefundNonce = formatInt(int(v))
	}
	switch v := raw["withdrawRequestedAt"].(type) {
	case float64:
		out.WithdrawRequestedAt = int(v)
	case int:
		out.WithdrawRequestedAt = v
	}
	return out
}

func readCorrectiveVoucherState(extra map[string]interface{}) *batchsettlement.BatchSettlementVoucherStateExtra {
	if extra == nil {
		return nil
	}
	raw, ok := extra["voucherState"].(map[string]interface{})
	if !ok {
		return nil
	}
	out := &batchsettlement.BatchSettlementVoucherStateExtra{}
	out.SignedMaxClaimable, _ = raw["signedMaxClaimable"].(string)
	out.Signature, _ = raw["signature"].(string)
	return out
}

func formatInt(n int) string {
	return big.NewInt(int64(n)).String()
}
